### ISSUE TITLE
Validating number ranges

### ISSUE THREAD

**Issue by emilbonnek** [fabian-hiller/valibot#20]
I couldn't find anyway to validate that a number is above/below some value or inside a range. Neither for inclusive ranges or exclusive ones.

In Zod this is done with the `gt` and `lt` for exclusive ranges and `gte` and `lte` for inclusive ranges. (Zod also has `positive`, `negative`, `nonpositive` and `nonnegative` for ranges bounded by `0`).
https://zod.dev/?id=numbers

(It is possible that this feature is already implemented in valibot and I just couldn't find it)

**Comment by fabian-hiller** [comment]
That works with `minRange` and `maxRange`:

```ts
import { maxRange, minRange, number } from "valibot";

const NumberSchema = number([minRange(0), maxRange(50)]);
```

**Comment by emilbonnek** [comment]
Thank you! are these inclusive ranges? (would 0 and 50 be valid in your example?)

I think `minRange`/`maxRange` has slightly unintuitive names for me. Maybe it won't matter once the documentation site is up but I believe the order of names I looked for was something like this before I posted the issue:
1. `gt`/`lt` - based on Zod's names
2. `min`/`max`
3. `minVal`/`maxVal` - (`minSize` and `maxSize` exists so I kind of expected a parity with the function names there)
4. `range`

**Comment by fabian-hiller** [comment]
Yes, 0 to 50 are valid values in this example. I know how this is implemented in other libraries, however, the approach I chose has a comprehensible reason.

Valibot is a completely modular library, which consists of many small functions. Some functions like `minRange` can be used in several contexts.

`minRange` uses the `<` operator to validate whether a value is at least as large as the specified value. If we only look at the source code of `minRange`, the name is exactly right.

`minRange` can now be used for validation in the pipeline of `string`, `number`, `bigint` and `date`. This approach reduces the bundle size and prevents duplicate code within the library.

Since there are several validation functions that validate a minimum value, for example also `minLength` and `minBytes`, the naming `min` is not possible. `minNumber` would be more suitable for `number`, but confusing for `string` and `date`.

However, I welcome alternative ideas to further improve the API. Feel free to share your thoughts here or create new issues if you have ideas or feedback on other areas of the library.

**Comment by kiliman** [comment]
The nice thing about `valibot` is that you can easily create your own validation functions:

```ts
export function gt<
  TInput extends string | number | bigint | Date,
  TRequirement extends TInput,
>(requirement: TRequirement, error?: string) {
  return (input: TInput, info: ValidateInfo) => {
    if (input <= requirement) {
      throw new ValiError([
        {
          validation: 'gt',
          origin: 'value',
          message: error || 'Invalid range',
          input,
          ...info,
        },
      ])
    }
    return input
  }
}

export function lte<
  TInput extends string | number | bigint | Date,
  TRequirement extends TInput,
>(requirement: TRequirement, error?: string) {
  return (input: TInput, info: ValidateInfo) => {
    if (input > requirement) {
      throw new ValiError([
        {
          validation: 'lte',
          origin: 'value',
          message: error || 'Invalid range',
          input,
          ...info,
        },
      ])
    }
    return input
  }
}
```
```ts
parse(number([gt(5), lte(80)]), 6)  // returns 6
parse(number([gt(5), lte(80)]), 5)  // throws
parse(number([gt(5), lte(80)]), 80) // returns 90
parse(number([gt(5), lte(80)]), 81)  // throws
```

**Comment by fabian-hiller** [comment]
Yes, or you can simply rename the exports:

```ts
import { maxRange as lte, minRange as gte, number } from "valibot";

const NumberSchema = number([lte(0), gte(50)]);
```

You can also easily export `minRange` yourself under several names like `minNumber` and `minDate`.

```ts
export { minRange as minNumber, minRange as minDate } from "valibot";
```

**Comment by emilbonnek** [comment]
I understand that I can just create my own validation function or even make named imports to have it the way that fits my personal preferences. But the concern I was raising is that `minRange` might be unintuitive for other developers as it was for me. This can make it harder to find the right function (which was the problem I had) and it can make the code harder to reason about (it's unrealistic that everyone will make the same named import - that only happens in Python land).

> `minRange` uses the `<` operator to validate whether a value is at least as large as the specified value. If we only look at the source code of [`minRange`](https://github.com/fabian-hiller/valibot/blob/main/library/src/validations/minRange/minRange.ts), the name is exactly right.

It might describe exactly what it does, but from the perspective of a user it might not be the most useful name. Try to read the following code and consider how a user might understand it:
- `string([minLength(x)])` -  validates that a `string`  has the minimum **length** of `x`.
- `map([minSize(x)])` -  validates that a `map` has the minimum **size** of `x`.
- `number([minRange(x)])` -  validates that a `number` has the minimum **range** of `x`.

But it doesn't make sense to say that a `number` or `date` has a "minimum range". You would maybe say a `number` has a minimum **value** of `x` or that a `date` has a minimum **date** of `x`. I am not sure if I can come up with a scenario where it would be make sense to use the comparison operator to validate a `string`, but I suspect that in such a scenario you wouldn't say that you are validating that the `string` has a minimum **range** of `x`.

> `minNumber` would be more suitable for `number`, but confusing for `string` and `date`.

I agree that minNumber would be bad for `string` and `date`, but I also think it would be bad for number. It doesn't make sense to say that the minimum **number** of a `number` is `x`. Also the code would haver "number" twice which is a bit redundant: `number([minNumber(x)])`.

> since there are several validation functions that validate a minimum value, for example also `minLength` and `minBytes`, the naming `min` is not possible.

I would argue neither of those functions validates minimum **value**. They validate the minimum **length** and the minimum **byte length** respectively. Why does it make renaming `minRange` to `min` impossible? Is there some technical reason I am missing here or do you think it will be confusing for developers using it? The typings would still prevent people from using the wrong one, in case they thought they could use `min` on an `array` or something.

> However, I welcome alternative ideas to further improve the API. Feel free to share your thoughts here or create new issues if you have ideas or feedback on other areas of the library.

**In summary:**
I think `min`/`max` would be preferable to `minRange`/`maxRange`. It intuitively makes more sense to me.
If it isn't possible for technical reasons I think it would be preferable to have separately named API's for `number` and `date` so we could use `minVal`/`maxVal` and `before`/`after` for them respectively (I understand that you like the modular approach where the same functions are used for multiple types, but I am just telling you what I find most sensible).

**Comment by fabian-hiller** [comment]
Thank you for taking the time to go into this in such detail. 💙 I agree with you that `minRange` is not intuitive.

What would you think about renaming `minRange` and `maxRange` to `minValue` and `maxValue`? That would make sense to me, especially after your input, since we validate the length with `minLength`, size with `minSize`, and then value with `minValue`.

**Comment by kiliman** [comment]
For what it's worth, the Web standard simply uses `min/max`

<img width="820" alt="image" src="https://github.com/fabian-hiller/valibot/assets/47168/58b67793-85db-47c3-968d-ac46c771a1e9">

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#:~:text=of%20autocomplete%20options-,max,-date%2C%20month

**Comment by fabian-hiller** [comment]
I see. My thought comes from the fact that besides `minLength` and `maxLength` there is also `length`. The same with size and bytes. To test an explicit value, we could also add `value`.

Also, when the other validation functions are taken into account, the approach with `minValue` makes the API more consistent for me. However, I also understand your thoughts.

**Comment by fabian-hiller** [comment]
I have now implemented it this way, but I am still open for discussion.

**Comment by lucaschultz** [comment]
I actually had the exact same issue as @emilbonnek. I think I too would prefer `min`/`max` for numbers as it is what I _intuitively_ reach for without checking the docs.

> For what it's worth, the Web standard simply uses `min/max`

This is a strong point for using `min`/`max` too.

Also, having written a few zod schemas involving dates, I think I would actually **prefer** separate `before`/`after` aliases for validating date ranges. This is because I often have to mentally process the concept that a _minimum date_ means all dates larger and therefore after it are valid. Having the function named ﻿`after` could help with easier understanding and a self-documenting API. 😊

**Comment by fabian-hiller** [comment]
Thank you for your feedback. I will think about it and look forward to other people's opinion about it.

**Comment by yamiteru** [comment]
Since this library is all about modularity I don't understand why we have these multipurpose functions. I'd think that I can validate different sizes and lengths like this:

```ts
number([gte(0), lte(50)]);
string([prop("length", [gt(10)]))])
```

**Comment by fabian-hiller** [comment]
These functions are not multipurpose. They each have a single and clearly defined task. If a function can be used in different places, it would not make sense to write the same code twice. Please explain in more detail what your statement refers to, so that I can better understand it and include it in the further development of the library.

**Comment by yamiteru** [comment]
> Yes, 0 to 50 are valid values in this example. I know how this is implemented in other libraries, however, the approach I chose has a comprehensible reason.
> 
> Valibot is a completely modular library, which consists of many small functions. Some functions like `minRange` can be used in several contexts.
> 
> `minRange` uses the `<` operator to validate whether a value is at least as large as the specified value. If we only look at the source code of [`minRange`](https://github.com/fabian-hiller/valibot/blob/main/library/src/validations/minRange/minRange.ts), the name is exactly right.
> 
> `minRange` can now be used for validation in the pipeline of `string`, `number`, `bigint` and `date`. This approach reduces the bundle size and prevents duplicate code within the library.
> 
> Since there are several validation functions that validate a minimum value, for example also `minLength` and `minBytes`, the naming `min` is not possible. `minNumber` would be more suitable for `number`, but confusing for `string` and `date`.
> 
> However, I welcome alternative ideas to further improve the API. Feel free to share your thoughts here or create new issues if you have ideas or feedback on other areas of the library.

@fabian-hiller here you say that this once function can be used in different contexts such as `string`, `number`, `bigint` and `date`. I don't think that's very "pure".

**Comment by kiliman** [comment]
@yamiteru, just remember this is @fabian-hiller's project. Even if you disagree, please keep it respectful.

You can always create your own custom functions. That's the reason I think this is better than existing libraries like `zod`. It is fully extensible while remaining slim.

**Comment by fabian-hiller** [comment]
@yamiteru [here](https://github.com/fabian-hiller/valibot/blob/main/library/src/validations/minValue/minValue.ts) is the source code. As you can see, it makes no difference if I allow only `string` or  also `number`, `bigint` and `Date` as input.

**Comment by yamiteru** [comment]
@kiliman How am I not respectful?
